libr/io/section.c

```diff
@@ -1,619 +1,745 @@
-/* radare2 - LGPL - Copyright 2008-2017 - pancake, nibble */
+/* radare2 - LGPL - Copyright 2017 - condret , alvarofe*/
 
-#include "r_io.h"
-// no link
-#include "r_cons.h"
+#include <r_io.h>
+#include <r_util.h>
+#include <sdb.h>
+#include <r_types.h>
+#include <stdio.h>
+#include <string.h>
+
+static void section_free(void *p) {
+	RIOSection *s = (RIOSection *) p;
+	if (s) {
+		free (s->name);
+		free (s);
+	}
+}
 
 R_API void r_io_section_init(RIO *io) {
-	io->next_section_id = 0;
-	io->enforce_rwx = 0; // do not enforce RWX section permissions by default
-	io->enforce_seek = 0; // do not limit seeks out of the file by default
-	io->sections = r_list_newf (r_io_section_free);
-	if (!io->sections) {
+	if (io && !io->sections) {
+		if (!(io->sections = ls_newf (section_free))) {
+			return;
+		}
+	}
+	io->sec_ids = r_id_pool_new (0, 0xffffffff);
+}
+
+R_API void r_io_section_fini(RIO *io) {
+	if (!io) {
 		return;
 	}
+	ls_free (io->sections);
+	r_id_pool_free (io->sec_ids);
+	io->sections = NULL;
+	io->sec_ids = NULL;
 }
 
-R_API RIOSection *r_io_section_get_name(RIO *io, const char *name) {
-	RListIter *iter;
-	RIOSection *s;
-	if (name) {
-		r_list_foreach (io->sections, iter, s) {
-			if (!strcmp (name, s->name)) {
-				return s;
-			}
+R_API int r_io_section_exists_for_id(RIO *io, ut32 id) {
+	SdbListIter *iter;
+	RIOSection *sec;
+	if (!io || !io->sections) {
+		return false;
+	}
+	ls_foreach (io->sections, iter, sec) {
+		if (sec->id == id) {
+			return true;
 		}
 	}
-	return NULL;
+	return false;
 }
 
-// update name and rwx, size is experimental
-static RIOSection *findMatching (RIO *io, ut64 paddr, ut64 vaddr, ut64 size, ut64 vsize, int rwx, const char *name) {
-	RListIter *iter;
-	RIOSection *s;
-	r_list_foreach (io->sections, iter, s) {
-		if (s->paddr != paddr) continue;
-		if (s->vaddr != vaddr) continue;
-#if 1
-		if (s->size != size) continue;
-		if (s->vsize != vsize) continue;
-#else
-		s->size = size;
-		s->vsize = vsize;
-#endif
-		s->flags = rwx;
-		if (name && strcmp (name, s->name)) {
-			s->name = strdup (name);
-		}
-		return s;
+// @deprecate
+RIOSection *_section_chk_dup(RIO *io, ut64 paddr, ut64 vaddr, ut64 size, ut64 vsize, int flags, const char *name, ut32 bin_id, int fd) {
+	RIOSection *sec;
+	SdbListIter *iter;
+	char sname[32];
+	if (!name) {
+		snprintf (sname, sizeof (sname) - 1, "section.0x016%"PFMT64x "", vaddr);
```

Comment:
low priority


For aesthetics, that's also why I prefer strscpy over `r_str_ncpy(dst, src, sizeof dst - 1);`

```diff
-		snprintf (sname, sizeof (sname) - 1, "section.0x016%"PFMT64x "", vaddr);
+		snprintf (sname, sizeof (sname), "section.0x016%"PFMT64x "", vaddr);
```